### DISCUSSION TITLE
Accelerated Mobile Pages (AMP)

### DISCUSSION THREAD

**Discussion by taufik-nurrohman** [mecha-cms/mecha#137] [Ideas]
These techniques are useful for those of you who want to make a mobile version of the page without having to sacrifice the existing desktop version of the page. You can have two versions of the page without having to make two web pages for each device separately.

One of the most popular mobile web page standards today is AMP, a web component framework to easily create user-first web experiences. Today, the AMP specifications may have changed, therefore I will not discuss it in detail here. You can learn the complete AMP specifications on the [official website](https://amp.dev/documentation).

Here I just want to focus on the methods you can use to tell Mecha that it should display the mobile version of the page if the conditions are met.

### Query String Method

This is the easiest method that you can implement right away by adding a mobile page specific conditional statement based on the presence of a particular query string in URL. In this case, the query string in question is `?amp=1`

``` php
State::set('is.mobile', !empty($_GET['amp']));
```

The code above will add a _mobile_ conditional feature that you can use in the layout file via the `$site` or `$state` object like so:

``` php
<!DOCTYPE html>
<?php if ($site->is('mobile')): ?>
  <html amp lang="en">
<?php else: ?>
  <html lang="en">
<?php endif; ?>
  <head> … </head>
  <body> … </body>
</html>
```

To activate the mobile version of the page, simply add an `?amp=1` query string in URL.

When working with different web page mode, helper functions to convert URL from the default version to the mobile version and vice-versa may be needed. The following are two helper functions you can use to work with query string based mobile web page:

``` php
From::_('mobileURL', static function (string $path) {
    if (!str_contains($path, '?')) {
        return $path;
    }
    $parts = explode('?', $path, 2);
    return $parts[0] . To::query(extend(From::query($parts[1] ?? ""), [
        'amp' => false
    ]));
});

To::_('mobileURL', static function (string $path) {
    if (!str_contains($path, '?')) {
        return $path . '?amp=1';
    }
    $parts = explode('?', $path, 2);
    return $parts[0] . To::query(extend(From::query($parts[1] ?? ""), [
        'amp' => '1'
    ]));
});
```

The following is an example of how to use it:

``` php
<?php if ($site->is('mobile')): ?>
  <a href="<?= To::mobileURL($page->url); ?>">
    This is a link to a page.
  </a>
<?php else: ?>
  <a href="<?= $page->url; ?>">
    This is a link to a page.
  </a>
<?php endif; ?>
```

### Domain Alias Method

Using a domain alias is also easy, but this requires you to hard-code the domain name in the source code:

``` php
State::set('is.mobile', 'm.example.com' === $url->host);
```
The following are two helper functions you can use to work with domain based mobile web page:

``` php
From::_('mobileURL', static function (string $path) use ($url) {
    // Do not touch relative URL
    if (!str_contains($path, '://')) {
        return $path;
    }
    $u = new URL($path);
    if ('m.example.com' !== $u->host) {
        return $path; // Do nothing!
    }
    $u->host = 'example.com';
    return $u->current;
});

To::_('mobileURL', static function (string $path) use ($url) {
    // Do not touch relative URL
    if (!str_contains($path, '://')) {
        return $path;
    }
    $u = new URL($path);
    if ('example.com' !== $u->host) {
        return $path; // Do nothing!
    }
    $u->host = 'm.example.com';
    return $u->current;
});
```

### Route Method

A little tricky, but still doable.

By adding these two route aliases, you will be able to have two types of URLs that will display the same page.

``` php
Hook::set('route', static function ($content, $path, $query, $hash) {
    if ($path && str_starts_with($path . '/', '/amp/')) {
        return Hook::fire('route', [$content, substr($path, 4), $query, $hash]);
    }
    return $content;
}, 19);
```

You can visit the page `http://127.0.0.1/article/lorem-ipsum` via `http://127.0.0.1/amp/article/lorem-ipsum` too, and they will both display the same page. Until you state that the mobile page conditional statement is met before the actual route execution.

``` php
Hook::set('route', static function ($content, $path, $query, $hash) {
    if ($path && str_starts_with($path . '/', '/amp/')) {
        State::set('is.mobile', true);
        return Hook::fire('route', [$content, substr($path, 4), $query, $hash]);
    }
    return $content;
}, 19);
```

The following are two helper functions you can use to work with route based mobile web page:

``` php
From::_('mobileURL', static function (string $path) use ($url) {
    // Empty URL
    if ("" === $path) {
        return $path;
    }
    // Relative URL
    if ('/amp' === $path) {
        return '/';
    }
    // Relative URL
    if (str_starts_with($path, '/amp/')) {
        return substr($path, 4);
    }
    // Full URL
    if ($url . '/amp' === $path) {
        return substr($path, 0, -4);
    }
    return strtr($path, [
        $url . '/amp/' => $url . '/'
    ]);
});

To::_('mobileURL', static function (string $path) use ($url) {
    // Empty URL
    if ("" === $path) {
        return '/amp';
    }
    // Relative URL
    if (str_starts_with($path, '/')) {
        return '/amp' . $path;
    }
    // Full URL
    if ($url . "" === $path) {
        return $path . '/amp';
    }
    return strtr($path, [
        $url . '/' => $url . '/amp/'
    ]);
});
```

## Hooking the URL, Globally

Applying helper functions to every URL inside the markup is a painstaking job. In addition, we also don&rsquo;t seem to be able to do much with the URLs in the page content. By imitating the way _link_ extension change URLs via the output buffer, we can make a simpler version of its hook to convert URLs only on `<a>` and `<form>` elements.

``` php
Hook::set('content', static function ($content) {
    if (!$content || !State::get('is.mobile')) {
        return $content;
    }
    if (str_contains($content, '</a>')) {
        $content = preg_replace_callback('/<a(\s[^>]*?)>/', static function ($m) {
            if (!str_contains($m[0], 'href=')) {
                return $m[0];
            }
            $e = new HTML($m[0]);
            $e['href'] = To::mobileURL($e['href']);
            return (string) $e;
        }, $content);
    }
    if (str_contains($content, '</form>')) {
        $content = preg_replace_callback('/<form(\s[^>]*?)>/', static function ($m) {
            if (!str_contains($m[0], 'action=')) {
                return $m[0];
            }
            $e = new HTML($m[0]);
            $e['action'] = To::mobileURL($e['action']);
            return (string) $e;
        }, $content);
    }
    return $content;
}, 0);
```

On a form of type `GET`, changing the action URL of the query string based mobile web page is useless.

The following is an example of a search form markup:

``` php
<form action="<?= $url; ?>/article" method="get">
  <p>
    <input name="query" type="text">
    <button type="submit">
      <?= i('Search'); ?>
    </button>
  </p>
</form>
```

Instead of appending query `?amp=1` to the action value, you can just add a hidden input element like this:

``` php
<form action="<?= $url; ?>/article" method="get">
  <p>
    <input name="query" type="text">
    <button type="submit">
      <?= i('Search'); ?>
    </button>
  </p>
  <?php if ($site->is('mobile')): ?>
    <input name="amp" type="hidden" value="1">
  <?php endif; ?>
</form>
```

Or, from the output buffer hook, you can insert the element like this:

``` php
$content = preg_replace_callback('/<form(\s[^>]*?)>/', static function ($m) {
    if (!str_contains($m[0], 'action=')) {
        return $m[0];
    }
    $e = new HTML($m[0]);
    if (!$e['method'] || 'get' === $e['method']) {
        return $m[0] . '<input name="amp" type="hidden" value="1">';
    }
    return $m[0];
}, $content);
```

It&rsquo;s not neat actually, but it works!

## Converting Native HTML Element into AMP Element

``` php
Hook::set('content', static function ($content) {
    if (!$content || !State::get('is.mobile')) {
        return $content;
    }
    if (str_contains($content, '</audio>')) {
        $content = preg_replace_callback('/<audio(\s[^>]*?)>[\s\S]*?<\/audio>/', static function ($m) {
            $e = new HTML($m[0]);
            $e[0] = 'amp-audio';
            return (string) $e;
        }, $content);
    }
    if (str_contains($content, '</iframe>')) {
        $content = preg_replace_callback('/<iframe(\s[^>]*?)>[\s\S]*?<\/iframe>/', static function ($m) {
            $e = new HTML($m[0]);
            $e[0] = 'amp-iframe';
            $e[1] = "";
            return (string) $e;
        }, $content);
    }
    if (str_contains($content, '<img ')) {
        $content = preg_replace_callback('/<img(\s[^>]*?)>/', static function ($m) {
            $e = new HTML($m[0]);
            $e[0] = 'amp-img';
            $e[1] = "";
            return (string) $e;
        }, $content);
    }
    if (str_contains($content, '</style>')) {
        $content = preg_replace_callback('/<style(\s[^>]*?)?>/', static function ($m) {
            if (str_contains($m[0], ' amp-boilerplate')) {
                return $m[0];
            }
            $e = new HTML($m[0]);
            $e['amp-custom'] = true;
            return (string) $e;
        }, $content);
    }
    if (str_contains($content, '</video>')) {
        $content = preg_replace_callback('/<video(\s[^>]*?)>[\s\S]*?<\/video>/', static function ($m) {
            $e = new HTML($m[0]);
            $e[0] = 'amp-video';
            return (string) $e;
        }, $content);
    }
    return $content;
}, 0);
```

**Comment by taufik-nurrohman** [comment]
In version 3.0.0, [`Route` class will be removed](https://github.com/mecha-cms/mecha/issues/152). As an alternative, you can use this technique which is free from context and so this will automatically apply to any of the child route hooks:

~~~ php
Hook::set('route', function($content, $path, $query, $hash) {
    if (str_starts_with($path, '/amp/')) {
        State::set('is.mobile', true);
        return Hook::fire('route', [$content, substr($path, 4), $query, $hash]);
    }
    return $content;
}, 0);
~~~

**Comment by yuliakbard** [comment]
kang, kode untuk amp yang route method, pada attribut class tag html menampilkan class is:error dan error:404, kalau pakai string query method gak ada class is:error dan error:404


kalau untuk `<link href="<?= $page->url ?>" rel="amphtml">` untuk `$page->url ` yang mendirect ke url route method pake yang mana kang? ane pakein ini `<?= To::mobileURL($page->url); ?>` gak jalan

**Reply by taufik-nurrohman** [reply]
Method `To::mobileURL()` harus dibuat dulu. Supaya nggak keluar error, set kondisional error menjadi `false`:

~~~ php
// < v3.0.0
Route::set('amp/*', function($any) {
    State::set('is.error', false);
    State::set('is.mobile', true);
    // ...
}, 19);
~~~

**Reply by yuliakbard** [reply]
yang kondisional error done kang.
yang Method `To::mobileURL()` yang ini kan kang?

``` php
From::_('mobileURL', function(string $path) use($url) {
    // Empty URL
    if ("" === $path) {
        return $path;
    }
    // Relative URL
    if ('/amp' === $path) {
        return '/';
    }
    // Relative URL
    if (str_starts_with($path, '/amp/')) {
        return substr($path, 4);
    }
    // Full URL
    return strtr($path, [
        $url . '/amp/' => $url . '/'
    ]);
});

To::_('mobileURL', function(string $path) use($url) {
    // Empty URL
    if ("" === $path) {
        return '/amp';
    }
    // Relative URL
    if (str_starts_with($path, '/')) {
        return '/amp' . $path;
    }
    // Full URL
    return strtr($path, [
        $url . '/' => $url . '/amp/'
    ]);
});
```

![Screenshot 2022-01-09 130540](https://user-images.githubusercontent.com/96925342/148671300-f7bb034f-48d6-4c9d-a4dd-7c691ae5c9b7.png)

![Screenshot 2022-01-09 130246](https://user-images.githubusercontent.com/96925342/148671207-af908ebc-ce39-43a5-b2f1-561913e2eec4.png)

tetap blank

**Reply by taufik-nurrohman** [reply]
Coba enable `DEBUG` mode.

**Reply by yuliakbard** [reply]
caranya gimana kang?

``` php
if(DEBUG === true){
masukin kode
}
```

seperti ini kah?

**Reply by taufik-nurrohman** [reply]
Di `.\index.php`:

~~~ php
define('DEBUG', true);
~~~

**Reply by yuliakbard** [reply]
udah kang, selanjutnya gimana kang?

**Reply by taufik-nurrohman** [reply]
Keluar pesan _error_ apa tidak?

**Reply by yuliakbard** [reply]
gak ada kang

**Reply by taufik-nurrohman** [reply]
Screenshot kode sumber.

**Reply by yuliakbard** [reply]
![Screenshot 2022-01-10 003637](https://user-images.githubusercontent.com/96925342/148693767-63777158-9a75-4a71-8948-8d3ab5368e58.png)
![Screenshot 2022-01-10 003610](https://user-images.githubusercontent.com/96925342/148693768-e482ecb7-1045-4fb9-9774-c99305c6763d.png)
![Screenshot 2022-01-10 003532](https://user-images.githubusercontent.com/96925342/148693769-c3eb2eb1-e6b2-4d0b-a5cc-540ef6757ae2.png)

**Reply by yuliakbard** [reply]
kang, saat saya backup dan jalankan di localhost computer, tereksekusi kang, ada output amp nya
![Screenshot 2022-01-10 004042](https://user-images.githubusercontent.com/96925342/148693911-035d4a80-8d94-43b0-b23e-548031ed95fa.png)
![Screenshot 2022-01-10 003946](https://user-images.githubusercontent.com/96925342/148693914-511f9885-068a-4c9e-99ae-a2487b4ba64a.png)

**Reply by yuliakbard** [reply]
Masalah AMP Fixed, PHP version nya ane naikkan jadi versi terbarunya baru jalan. versi PHP 8.1

**Comment by yuliakbard** [comment]
`<link href="<?= $page->url ?>" rel="amphtml">`

Kang, bagaimana caranya agar html diatas hanya muncul pada halaman selain path `/amp`?
penulisan kondisionalnya seperti apa?

**Reply by taufik-nurrohman** [reply]
Kondisionalnya ya itu mas yang udah kamu bikin sendiri:

~~~ php
<?php if (!$site->is('mobile')): ?>
  <link href="<?= $page->url ?>" rel="amphtml">
<?php endif; ?>
~~~

**Comment by yuliakbard** [comment]


**Reply by taufik-nurrohman** [reply]
Dibikin topik baru aja mas.
